### PR TITLE
Allow 0 as Retry_Limit to disable retrying 

### LINKED ISSUE #3190: Add Option to Disable retries (or allow 0 for retries)
**Is your feature request related to a problem? Please describe.**

In some cases, logs are coming in at such high throughput that even trying to retry is not ideal. May be buffering options are limited.

I have a customer who would prefer to be able to set retries to be 0, which is not possible today: https://docs.fluentbit.io/manual/administration/scheduling-and-retries

It should be super simple to enable this.

### PR DESCRIPTION
**Pull request by nokute78**
Fixes #3190 

This PR allows `Retry_Limit 0` and it disables retrying.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:
- [X] Example configuration file for the change
- [X] Debug log output from testing the change

- [X] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

**Documentation**

- [ ] Documentation required for this feature

## Example Configuration

```
[INPUT]
    Name dummy
    Samples 1

[OUTPUT]
    Name es
    Retry_Limit 0
```

## Debug output

Stop Elasticsearch before testing.
```
$ ../bin/fluent-bit -c a.conf 
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/03/13 12:52:41] [ info] [engine] started (pid=100875)
[2021/03/13 12:52:41] [ info] [storage] version=1.1.1, initializing...
[2021/03/13 12:52:41] [ info] [storage] in-memory
[2021/03/13 12:52:41] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/03/13 12:52:41] [ info] [sp] stream processor started
[2021/03/13 12:52:46] [error] [src/flb_http_client.c:1156 errno=32] Broken pipe
[2021/03/13 12:52:46] [ warn] [output:es:es.0] http_do=-1 URI=/_bulk
[2021/03/13 12:52:46] [ info] [engine] chunk '100875-1615607562.413281173.flb' is not retried (no retry config): task_id=0, input=dummy.0 > output=es.0 (out_id=0)
^C[2021/03/13 12:53:17] [engine] caught signal (SIGINT)
[2021/03/13 12:53:17] [ warn] [engine] service will stop in 5 seconds
[2021/03/13 12:53:21] [ info] [engine] service stopped
```

## Valgrind output

```
$ valgrind ../bin/fluent-bit -c a.conf 
==100878== Memcheck, a memory error detector
==100878== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==100878== Using Valgrind-3.15.0 and LibVEX; rerun with -h for copyright info
==100878== Command: ../bin/fluent-bit -c a.conf
==100878== 
Fluent Bit v1.8.0
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/03/13 12:53:48] [ info] [engine] started (pid=100878)
[2021/03/13 12:53:48] [ info] [storage] version=1.1.1, initializing...
[2021/03/13 12:53:48] [ info] [storage] in-memory
[2021/03/13 12:53:48] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/03/13 12:53:48] [ info] [sp] stream processor started
==100878== Warning: client switching stacks?  SP change: 0x57e48b8 --> 0x4c6bd20
==100878==          to suppress, use: --max-stackframe=12028824 or greater
==100878== Warning: client switching stacks?  SP change: 0x4c6bc98 --> 0x57e48b8
==100878==          to suppress, use: --max-stackframe=12028960 or greater
==100878== Warning: client switching stacks?  SP change: 0x57e48b8 --> 0x4c6bc98
==100878==          to suppress, use: --max-stackframe=12028960 or greater
==100878==          further instances of this message will not be shown.
[2021/03/13 12:53:52] [error] [src/flb_http_client.c:1156 errno=32] Broken pipe
[2021/03/13 12:53:52] [ warn] [output:es:es.0] http_do=-1 URI=/_bulk
[2021/03/13 12:53:52] [ info] [engine] chunk '100878-1615607628.446813184.flb' is not retried (no retry config): task_id=0, input=dummy.0 > output=es.0 (out_id=0)
^C[2021/03/13 12:54:32] [engine] caught signal (SIGINT)
[2021/03/13 12:54:32] [ warn] [engine] service will stop in 5 seconds
[2021/03/13 12:54:36] [ info] [engine] service stopped
==100878== 
==100878== HEAP SUMMARY:
==100878==     in use at exit: 0 bytes in 0 blocks
==100878==   total heap usage: 534 allocs, 534 frees, 1,652,774 bytes allocated
==100878== 
==100878== All heap blocks were freed -- no leaks are possible
==100878== 
==100878== For lists of detected and suppressed errors, rerun with: -s
==100878== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

File: src/flb_output.c
Comment (PettitWesley) [inline src/flb_output.c:613]: What user input leads to this else case being run? I'm trying to follow the logic and I don't get it..

Comment (nokute78) [reply src/flb_output.c:613]: We can set via command line. `-p "Retry_Limit="`
e.g. `$ ../bin/fluent-bit -i dummy -o es -p "Retry_Limit=" --sosreport`

Without this patch, if we set retry_limit=0, fluent-bit will retry again and again.

Comment (nokute78) [reply src/flb_output.c:613]: (By the way, I think this value should be 1 (=default value) at this else case.)

Comment (PettitWesley) [reply src/flb_output.c:613]: > (By the way, I think this value should be 1 (=default value) at this else case.)

I agree

Comment (nokute78) [reply src/flb_output.c]: Thank you.
I've pushed a commit to set default value 1.

**Review by PettitWesley — APPROVED** [review]

**Comment by PettitWesley** [comment]
@edsiper Can we merge this?

**Comment by PettitWesley** [comment]
@nokute78 Can you cherry-pick these commits onto the 1.7 branch so that we can get this change released in the next 1.7 patch release?

**Comment by nokute78** [comment]
@PettitWesley No I can't.
I've never released fluent-bit..

**Comment by PettitWesley** [comment]
@nokute78 Yes you can! Lol I just meant open a pull request against the 1.7 branch. I can merge it for you. I can't do releases either but if we merge it then it will go out in the next release.

**Comment by nokute78** [comment]
@PettitWesley Sure.
I open new pull request #3261 for 1.7 branch.

**Comment by edsiper** [comment]
@nokute78 

can we use some words instead of numbers as the suggested values ?, e.g:

- the current context per configuration variable name is ```retry_limit```, which aims to specify "what's the limit of retries ?"

if we set "0" from a reading perspective might mean "there is no limit", so I propose the following values and behaviors:

| value | description / behavior |
|--|--|
| 1-9 | specify the maximum limit of retries that can be done. The value is any integer value greater than zero. |
| ```no_retries``` | do not attempt to perform any retry |
| ```no_limits``` or ```false``` | retry forever until it succeeds |

my concern is the user readability of the expected behavior more than the technical implementation.

@PettitWesley comments?

**Comment by nokute78** [comment]
@edsiper  Thank you for comment. I agree.
Or another way is to create new option.

Fluentd supports so many configuration parameters for retrying.
It means that there are so many requests for controlling retrying.
https://docs.fluentd.org/output#control-retrying

Currently `retry_limit` supports functionalities  `retry_forever` and `retry_max_times` of Fluentd.
To extract the meaning of `retry_limit` may not handle the so many use cases.

**Comment by edsiper** [comment]
I would avoid adding more options if the original configuration property can handle all situations

**Comment by edsiper** [comment]
this PR is on hold until the logic of the PR is changed as suggested on https://github.com/fluent/fluent-bit/pull/3219#issuecomment-803429303

**Comment by edsiper** [comment]
@niedbalski can you check pls why the message check is failing?

**Comment by nokute78** [comment]
> can you check pls why the message check is failing?

https://github.com/fluent/fluent-bit/blob/master/.github/workflows/check-commit.yaml#L17
This pattern doesn't allow `/`, so the commit message 
`output: define unlimited/none retry configurtion constant`
is invalid.

```ruby
$ irb
irb(main):001:0> str = "output: define unlimited/none retry configurtion constant"
irb(main):002:0> p str.match /^[a-z\-_]+\:[ ]{0,1}[a-z]+[a-zA-Z0-9 \-\.\:_]+$/
nil
=> nil
irb(main):003:0> p str.match /^[a-z\-_]+\:[ ]{0,1}[a-z]+[a-zA-Z0-9 \/\-\.\:_]+$/
#<MatchData "output: define unlimited/none retry configurtion constant">
=> #<MatchData "output: define unlimited/none retry configurtion constant">

```

**Comment by nokute78** [comment]
I updated this PR to follow https://github.com/fluent/fluent-bit/pull/3219#issuecomment-803429303

**Comment by nokute78** [comment]
@PettitWesley Would you comment about https://github.com/fluent/fluent-bit/pull/3219#issuecomment-803429303 ? 
I already pushed the patch to follow that comment.

**Comment by PettitWesley** [comment]
@nokute78 @edsiper 

> if we set "0" from a reading perspective might mean "there is no limit", 
> I would avoid adding more options if the original configuration property can handle all situations


I agree with Eduardo's reasoning, and his proposal. I'd use `no_retries` to mean don't retry at all- and I would use `false` for no limit.

**Comment by nokute78** [comment]
@PettitWesley Thank you for comment.

@edsiper Can we merge this ?
- This PR support your proposal https://github.com/fluent/fluent-bit/pull/3219#issuecomment-803429303
- `publish the docker images for PR` failed. I think it doesn't relate this PR by [log](https://github.com/fluent/fluent-bit/runs/2197210043?check_suite_focus=true)
- Commit message check failed, but [current master pattern](https://github.com/fluent/fluent-bit/commit/223b4bc55da9a3f326fa0a8f948b752ac31eef09) will not fail.
```ruby
irb(main):001:0> str = 'output: define unlimited/none retry configurtion constan
t'
irb(main):002:0> str.match /^[a-z0-9A-Z\-_\s\,\.\/]+\:[ ]{0,1}[a-zA-Z]+[a-zA-Z0-
9 \-\.\:_\#\(\)=\/\'\"\,><\+\[\]\!\*\\]+$/
=> #<MatchData "output: define unlimited/none retry configurtion constant">
```

**Comment by JeffLuoo** [comment]
Hi @edsiper Could you please take a look at this PR and to see if we can merge it?

**Comment by edsiper** [comment]
@nokute78 we are good to go, please just do the minor changes to pass the CI

**Comment by nokute78** [comment]
@edsiper CI status is https://github.com/fluent/fluent-bit/pull/3219#issuecomment-812215314

There are two errors, but one is not related this PR and the other should be fixed latest pattern.

====

- `publish the docker images for PR` failed. I think it doesn't relate this PR by [log](https://github.com/fluent/fluent-bit/runs/2197210043?check_suite_focus=true)
- Commit message check failed, but [current master pattern](https://github.com/fluent/fluent-bit/commit/223b4bc55da9a3f326fa0a8f948b752ac31eef09) will not fail.
```ruby
irb(main):001:0> str = 'output: define unlimited/none retry configurtion constan
t'
irb(main):002:0> str.match /^[a-z0-9A-Z\-_\s\,\.\/]+\:[ ]{0,1}[a-zA-Z]+[a-zA-Z0-
9 \-\.\:_\#\(\)=\/\'\"\,><\+\[\]\!\*\\]+$/
=> #<MatchData "output: define unlimited/none retry configurtion constant">
```

**Comment by nokute78** [comment]
Note: Commit message regex pattern is not updated by re-run CI job.

**Comment by nokute78** [comment]
`Run integration tests for PR` is failed at `Wait for docker build to success`.
It took 1hour, it means time out happened.

However previous step `Build jakejarvis/wait-action@master` was success.

@niedbalski 
Is it related following issue ? It reports wrong sha issue.
https://github.com/fountainhead/action-wait-for-check/issues/8

**Comment by nokute78** [comment]
Only `Commit Message Check` failed and code owner review required.
What should I do ?

**Comment by PettitWesley** [comment]
Looks like the check wasn't run on your latest commits: https://github.com/fluent/fluent-bit/pull/3219/checks?check_run_id=2311012858

**Comment by nokute78** [comment]
@PettitWesley Thank you for comment.
"Check Commit Message" will not be success.
Because its regex pattern is not updated by re-running CI. The pattern is different from current master...

I tested by hand and it was success https://github.com/fluent/fluent-bit/pull/3219#issuecomment-817070894

**Comment by PettitWesley** [comment]
> @PettitWesley Thank you for comment.
"Check Commit Message" will not be success.
Because its regex pattern is not updated by re-running CI. The pattern is different from current master...

> I tested by hand and it was success #3219 (comment)


I think only @edsiper or @agup006 Can fix this problem

**Comment by nokute78** [comment]
@PettitWesley I think so.

@edsiper @agup006 @niedbalski What should I do?

**Comment by agup006** [comment]
Merged for now, @niedbalski maybe we need the Commit message check to rerun on fix?

**Comment by nokute78** [comment]
I removed draft status from document repo.
https://github.com/fluent/fluent-bit-docs/pull/510

@agup006 @niedbalski Thank you.

FYI: CI was succeeded when it was merged master.
https://github.com/fluent/fluent-bit/tree/a9801ab0a9e464c3098be868b52057330019120a

**Pull request merged**
